internal/xds/translator/session_persistence.go

```diff
@@ -0,0 +1,140 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package translator
+
+import (
+	"errors"
+	"strings"
+
+	routev3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	hcmv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	cookiev3 "github.com/envoyproxy/go-control-plane/envoy/extensions/http/stateful_session/cookie/v3"
+	headerv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/http/stateful_session/header/v3"
+	httpv3 "github.com/envoyproxy/go-control-plane/envoy/type/http/v3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"github.com/envoyproxy/gateway/internal/ir"
+	"github.com/envoyproxy/gateway/internal/xds/types"
+)
+
+const (
+	sessionPersistenceFilter = "envoy.filters.http.stateful_session"
+)
+
+type sessionPersistence struct{}
+
+func init() {
+	registerHTTPFilter(&sessionPersistence{})
+}
+
+var _ httpFilter = &sessionPersistence{}
+
+// patchHCM patches the HttpConnectionManager with the filter.
+// Note: this method may be called multiple times for the same filter, please
+// make sure to avoid duplicate additions of the same filter.
+func (s *sessionPersistence) patchHCM(mgr *hcmv3.HttpConnectionManager, irListener *ir.HTTPListener) error {
+	if mgr == nil {
+		return errors.New("hcm is nil")
+	}
+
+	if irListener == nil {
+		return errors.New("ir listener is nil")
+	}
+
+	// Return early if filter already exists.
+	for _, f := range mgr.HttpFilters {
+		if f.Name == sessionPersistenceFilter {
+			return nil
+		}
+	}
+
+	for _, route := range irListener.Routes {
+		sp := route.SessionPersistence
+		if sp == nil {
+			continue
+		}
+
+		var cfg proto.Message
+		switch {
+		case sp.Cookie != nil:
+			cfg = &cookiev3.CookieBasedSessionState{
+				Cookie: &httpv3.Cookie{
+					Name: sp.SessionName,
+					Path: routePathToCookiePath(route.PathMatch),
+					Ttl:  durationpb.New(sp.Cookie.TTL.Duration),
+				},
+			}
+		case sp.Header != nil:
+			cfg = &headerv3.HeaderBasedSessionState{
+				Name: sp.SessionName,
+			}
+		}
+
+		cfgAny, err := anypb.New(cfg)
+		if err != nil {
+			return err
+		}
+
+		mgr.HttpFilters = append(mgr.HttpFilters, &hcmv3.HttpFilter{
```

Comment:
You need to use different name for stateful session filters on different routes, disable them on the listener level, and enable them on their respective routes. You can find an example in OIDC.go https://github.com/envoyproxy/gateway/blob/main/internal/xds/translator/oidc.go